src/compiler/checker.ts

```diff
@@ -36957,7 +36967,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // There is no point in doing an assignability check if the function
         // has no explicit return type because the return type is directly computed
         // from the yield expressions.
-        const returnType = getReturnTypeFromAnnotation(func);
+        let returnType = getReturnTypeFromAnnotation(func);
+        if (returnType && returnType.flags & TypeFlags.Union) {
+            returnType = filterType(returnType, t => checkGeneratorInstantiationAssignabilityToReturnType(t, functionFlags, /*errorNode*/ undefined));
+        }

```

Comment:
Is this necessary? I think it makes sense to be more forgiving for contextual types, but if you write:
```ts
function * f(): Generator<A, B, C> | string {}
```
then you'll probably have other issues with assignability. I'm not sure we should filter the return type of an explicit type annotation.

Is this to support #41428? If so, I'm not sure we should try to handle a case of

```ts
function * f(): Generator<A> | AsyncGenerator<A> {}
```

since a normal generator is incapable of producing an async generator.

Reply:
> Is this to support https://github.com/microsoft/TypeScript/issues/41428? If so, I'm not sure we should try to handle a case of

Yes, it was added in this commit: ["Filter annotated return type within `checkYieldExpression`"](https://github.com/microsoft/TypeScript/pull/51196/commits/dfe955a861eec22029dc12bd5fc2f0db881ede12)

> since a normal generator is incapable of producing an async generator.

That is true, I just assumed that this might be wanted to handle mixed unions. The `function*` is just sugar and with regular functions, we are free to assign wider return types than what is required to be covered. So, from my PoV, this was meant as a convenience - it's more about type reusability than about actually wanting to write a union like this directly as the return type annotation, usually it would come from an alias.